mobile/library/jni/jni_impl.cc

```diff
@@ -827,24 +824,58 @@ extern "C" JNIEXPORT jlong JNICALL Java_io_envoyproxy_envoymobile_engine_JniLibr
 }
 
 extern "C" JNIEXPORT jint JNICALL Java_io_envoyproxy_envoymobile_engine_JniLibrary_startStream(
-    JNIEnv* env, jclass, jlong engine_handle, jlong stream_handle, jobject j_context,
+    JNIEnv* env, jclass, jlong engine_handle, jlong stream_handle, jobject java_stream_callbacks,
     jboolean explicit_flow_control) {
-  // TODO: To be truly safe we may need stronger guarantees of operation ordering on this ref.
-  jobject retained_context = env->NewGlobalRef(j_context);
-  envoy_http_callbacks native_callbacks = {jvm_on_response_headers,
-                                           jvm_on_response_data,
-                                           jvm_on_metadata,
-                                           jvm_on_response_trailers,
-                                           jvm_on_error,
-                                           jvm_on_complete,
-                                           jvm_on_cancel,
-                                           jvm_on_send_window_available,
-                                           retained_context};
+  Envoy::JNI::JniHelper jni_helper(env);
+  auto java_stream_callbacks_global_ref = jni_helper.newGlobalRef(java_stream_callbacks).release();
   auto engine = reinterpret_cast<Envoy::InternalEngine*>(engine_handle);
+  Envoy::EnvoyStreamCallbacks stream_callbacks;
+  // TODO(fredyw): Rewrite the whole implementation to not use the bridge data structures.
+  stream_callbacks.on_headers_ =
+      [java_stream_callbacks_global_ref](const Envoy::Http::ResponseHeaderMap& headers,
+                                         bool end_stream, envoy_stream_intel stream_intel) {
+        envoy_headers bridge_headers = Envoy::Http::Utility::toBridgeHeaders(headers);
+        jvm_on_response_headers(bridge_headers, end_stream, stream_intel,
+                                java_stream_callbacks_global_ref);
+      };
+  stream_callbacks.on_data_ = [java_stream_callbacks_global_ref](
+                                  const Envoy::Buffer::Instance& buffer, uint64_t length,
+                                  bool end_stream, envoy_stream_intel stream_intel) {
+    envoy_data bridge_data = Envoy::Data::Utility::toBridgeData(buffer, length);
```

Comment:
question: will this bridge eventually be removed in favor of using the Buffer::Instance (or ResponseHeaderMap for on_headers_) directly?

Reply:
Yeah, I have a TODO above to rewrite the entire JNI implementation. It will done separately, though since this PR is already big enough.